clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java

```diff
@@ -71,12 +71,12 @@
     /**
      * @see KafkaConsumer#commit(Map, CommitType)
      */
-    public void commit(Map<TopicPartition, Long> offsets, CommitType commitType);
+    public void commit(Map<TopicPartition, OffsetMetadata> offsets, CommitType commitType);
```

Comment:
Hmm, not sure I like that API. Having a map for the offsets and then requiring all TPs to use the same metadata is odd. And it seems pretty annoying if you want to do async and get a callback when the commit completes -- now you have to manage waiting on all the callbacks to complete as well. Also, splitting the commit across a bunch of separate requests isn't ideal.

Another alternative is to add this variant:

```
public void commit(Map<TopicPartition, OffsetAndMetadata> offsets, CommitType commitType, ConsumerCommitCallback callback);
```

but leave the existing `Map<TopicPartition, Long>` versions. If you want to include metadata, you *must* provide all the other parameters, which avoids introducing 2x the number of APIs and is reasonable since using metadata is an advanced feature. This has the same impact as the `String commitMessage` variant since I assume that would be an additional method since most users do not want it. So both would add 1 method, but this version does everything in one round trip, only has one callback, etc.

Copycat actually doesn't have a use case for the metadata currently, so I unfortunately don't have a good example use case for any of this currently.